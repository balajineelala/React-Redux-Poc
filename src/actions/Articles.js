import fetchArticles from "../services/Articles.js";

export function getArticles() {
  return async function getArticlesThunk(dispatch) {
    try {
      const response = await fetchArticles();
      dispatch(getArticlesSuccess(response.data));
    } catch (error) {
      dispatch(getArticlesFailure(error));
    }
  };
}

export const GET_ARTICLES_FAILURE = "GET_ARTICLES_FAILURE";

export function getArticlesFailure(error) {
  return {
    type: GET_ARTICLES_FAILURE,
    error,
  };
}

export const GET_ARTICLES_SUCCESS = "GET_ARTICLES_SUCCESS";

export function getArticlesSuccess(articlesList) {
  return {
    type: GET_ARTICLES_SUCCESS,
    articlesList,
  };
}
