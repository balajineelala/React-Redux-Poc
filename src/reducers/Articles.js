import {
  GET_ARTICLES_SUCCESS,
  GET_ARTICLES_FAILURE,
} from "../actions/Articles.js";

const initialState = {
  articlesList: [],
};

function Articles(state = initialState, action) {
  switch (action.type) {
    case GET_ARTICLES_SUCCESS:
      return {
        ...state,
        articlesList: action.articlesList,
      };
    case GET_ARTICLES_FAILURE:
      return {
        error: action.error,
      };
    default:
      return state;
  }
}

export default Articles;
