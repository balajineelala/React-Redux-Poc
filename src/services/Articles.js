import axios from "axios";

export default function fetchArticles() {
  return axios.get("https://jsonplaceholder.typicode.com/posts");
}
