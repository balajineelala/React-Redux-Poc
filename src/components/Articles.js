import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { getArticles } from "../actions/Articles";
import ArticlesList from "./ArticlesList";

function Articles() {
  const { articlesList } = useSelector((state) => state.articles);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getArticles());
  }, [dispatch]);

  return (
    <div>
      <ArticlesList articles={articlesList} />
    </div>
  );
}

export default React.memo(Articles);
