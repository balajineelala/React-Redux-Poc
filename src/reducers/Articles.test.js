import Articles from "./Articles";
import {
  GET_ARTICLES_SUCCESS,
  GET_ARTICLES_FAILURE,
} from "../actions/Articles.js";

describe("Articles reducer", () => {
  const initialState = {
    articlesList: [],
  };

  it("Articles reducer returns the initial state when an action type is not passed", () => {
    const reducer = Articles(undefined, {});
    expect(reducer).toEqual(initialState);
    expect(reducer).toMatchSnapshot();
  });

  it("Articles reducer handles GET_ARTICLES_SUCCESS as expected", () => {
    const reducer = Articles(initialState, {
      type: GET_ARTICLES_SUCCESS,
      articlesList: [
        {
          id: 1,
          title: "test title",
          userId: "test userId",
          body: "test body",
        },
      ],
    });

    expect(reducer).toEqual({
      articlesList: [
        {
          id: 1,
          title: "test title",
          userId: "test userId",
          body: "test body",
        },
      ],
    });
    expect(reducer).toMatchSnapshot();
  });

  it("Articles reducer handles GET_ARTICLES_FAILURE as expected", () => {
    const reducer = Articles(initialState, {
      type: GET_ARTICLES_FAILURE,
      error: { error: "test error" },
    });

    expect(reducer).toEqual({ error: { error: "test error" } });
    expect(reducer).toMatchSnapshot();
  });
});
