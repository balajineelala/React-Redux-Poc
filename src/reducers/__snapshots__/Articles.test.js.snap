// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles reducer Articles reducer handles GET_ARTICLES_FAILURE as expected 1`] = `
Object {
  "error": Object {
    "error": "test error",
  },
}
`;

exports[`Articles reducer Articles reducer handles GET_ARTICLES_SUCCESS as expected 1`] = `
Object {
  "articlesList": Array [
    Object {
      "body": "test body",
      "id": 1,
      "title": "test title",
      "userId": "test userId",
    },
  ],
}
`;

exports[`Articles reducer Articles reducer returns the initial state when an action type is not passed 1`] = `
Object {
  "articlesList": Array [],
}
`;
